File: src/services/auditService.js
import { apiClient, IS_MOCK_FALLBACK } from './api';
import { mockAuditLogs } from '../mock/auditLogs';

const isSandboxModeActive = () => {
  try {
    return localStorage.getItem('cee_is_sandbox') === 'true';
  } catch {
    return false;
  }
};

let sandboxAuditLogsState = [...mockAuditLogs];

const getGenuineAuditLogs = () => {
  try {
    const raw = localStorage.getItem('cee_genuine_audit');
    return raw ? JSON.parse(raw) : [];
  } catch {
    return [];
  }
};

const saveGenuineAuditLogs = (list) => {
  try {
    localStorage.setItem('cee_genuine_audit', JSON.stringify(list));
  } catch (err) {
    console.error('Failed to persist genuine audit logs:', err);
  }
};

export const auditService = {
  async getAuditLogs(filters = {}) {
    try {
      if (!IS_MOCK_FALLBACK) {
        const response = await apiClient.get('/audit', { params: filters });
        return response.data;
      }
    } catch (err) {
      // Fallback
    }

    const sourceList = isSandboxModeActive() ? sandboxAuditLogsState : getGenuineAuditLogs();

    return sourceList.filter((log) => {
      if (filters.event && filters.event !== 'ALL' && log.event !== filters.event) {
        return false;
      }
      if (filters.organization && filters.organization !== 'ALL' && !log.organization.includes(filters.organization)) {
        return false;
      }
      if (filters.search) {
        const q = filters.search.toLowerCase();
        return (
          log.id.toLowerCase().includes(q) ||
          log.evidenceId.toLowerCase().includes(q) ||
          log.actor.toLowerCase().includes(q) ||
          log.event.toLowerCase().includes(q) ||
          log.details.toLowerCase().includes(q) ||
          log.reference.toLowerCase().includes(q)
        );
      }
      return true;
    });
  },

  async logEvent(logPayload) {
    const newLog = {
      id: `AUD-${Math.floor(1000 + Math.random() * 9000)}`,
      timestamp: new Date().toISOString().replace('T', ' ').substring(0, 19) + ' UTC',
      evidenceId: logPayload.evidenceId || 'N/A',
      event: logPayload.event || 'SYSTEM_ACTION',
      actor: logPayload.actor || 'system',
      organization: logPayload.organization || 'Local Node',
      details: logPayload.details || 'Cryptographic audit log recorded',
      reference: logPayload.reference || '0x' + Array.from({length: 40}, () => Math.floor(Math.random()*16).toString(16)).join(''),
      verification: 'VERIFIED'
    };

    if (isSandboxModeActive()) {
      sandboxAuditLogsState.unshift(newLog);
    } else {
      const current = getGenuineAuditLogs();
      current.unshift(newLog);
      saveGenuineAuditLogs(current);
    }

    return newLog;
  }
};

File: backend/app/services/audit_service.py
from sqlalchemy.orm import Session
from app.models.audit_log import AuditLog
from app.schemas.audit import AuditLogResponse

class AuditService:
    @staticmethod
    def get_logs(db: Session, event: str = None, organization: str = None, search: str = None):
        query = db.query(AuditLog)
        if event and event != 'ALL':
            query = query.filter(AuditLog.event == event)
        if organization and organization != 'ALL':
            query = query.filter(AuditLog.organization.ilike(f"%{organization}%"))
        if search:
            search = search.lower()
            query = query.filter(
                (AuditLog.id.ilike(f"%{search}%")) |
                (AuditLog.evidence_id.ilike(f"%{search}%")) |
                (AuditLog.actor.ilike(f"%{search}%")) |
                (AuditLog.event.ilike(f"%{search}%")) |
                (AuditLog.details.ilike(f"%{search}%"))
            )
            
        results = query.order_by(AuditLog.timestamp.desc()).all()
        return [AuditService._format_response(r) for r in results]

    @staticmethod
    def _format_response(a: AuditLog) -> AuditLogResponse:
        return AuditLogResponse(
            id=a.id,
            timestamp=a.timestamp.strftime('%Y-%m-%d %H:%M:%S UTC') if a.timestamp else "",
            event=a.event,
            actor=a.actor,
            organization=a.organization,
            evidenceId=a.evidence_id,
            eventId=a.event_id,
            verification=a.verification,
            reference=a.reference,
            details=a.details
        )
